File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/ElementCollection.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.ChangeType;
import com.hello.uxml.tools.framework.events.CollectionChangedEvent;

import java.util.ArrayList;
import java.util.List;

/**
 * Implements an observable collection.
 *
 * @author ferhat
 */
public class ElementCollection extends UxmlElement {

  /** internal list of items */
  private List<Object> items;
  /** instance of event arguments that is reused across multiple collection changes */
  private CollectionChangedEvent cachedChangeEvent;

  public ElementCollection() {
    super();
  }

  /**
   * Returns size of collection.
   */
  public int size() {
    return (items != null) ? items.size() : 0;
  }

  /**
   * Adds item to collection.
   */
  @ContentNode
  public void add(Object item) {
    if (items == null) {
      items = new ArrayList<Object>();
    }
    items.add(item);
    raiseChangedEvent(ChangeType.Add, items.size() - 1, 1);
  }

  /**
   * Inserts item into collection.
   * @param index Insertion index.
   * @param item object to insert into collection.
   */
  public void insert(int index, Object item) {
    if (items == null) {
      items = new ArrayList<Object>();
    }
    items.add(index, item);
    raiseChangedEvent(ChangeType.Add, index, 1);
  }

  /**
   * Removes item from collection.
   */
  public void remove(Object item) {
    if (items != null) {
      int index = items.indexOf(item);
      if (index != -1) {
        items.remove(index);
        raiseChangedEvent(ChangeType.Remove, index, 1);
      }
    }
  }

  /**
   * Removes item at index from collection.
   */
  public void removeAt(int index) {
    if ((index >= 0) && (items != null) && (index < size())) {
      items.remove(index);
      raiseChangedEvent(ChangeType.Remove, index, 1);
    }
  }

  /**
   * Removes all items from collection.
   */
  public void clear() {
    if ((items != null) && (items.size() != 0)) {
      int count = items.size();
      items.clear();
      raiseChangedEvent(ChangeType.Remove, 0, count);
    }
  }

  /**
   * Returns item at index.
   */
  public Object get(int index) {
    return items.get(index);
  }

  private void raiseChangedEvent(ChangeType type, int index, int count) {
    if (cachedChangeEvent == null) {
      cachedChangeEvent = new CollectionChangedEvent(this, type, index, count);
    } else {
      cachedChangeEvent.setType(type);
      cachedChangeEvent.setIndex(index);
      cachedChangeEvent.setCount(count);
    }
    this.notifyListeners(CollectionChangedEvent.changeEvent, cachedChangeEvent);
  }
}

File: framework/lib/element_collection.dart
part of uxml;

/**
 * Implements an observable collection.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class ElementCollection extends UxmlElement {
  List<Object> _items;
  static const int _NUM_ITEMS_CHANGED_ONE_ITEM = 1;
  static const int _START_CHANGE_INDEX = 0;

  /** Collection change event definition. */
  static EventDef changedEvent;
  static void initElementCollection() {
    changedEvent = CollectionChangedEvent.eventDef;
  }

  ElementCollection() : super() {
  }

  /**
   * Returns size of collection.
   */
  int get length => _items == null ? 0 : _items.length;

  /**
   * Adds items to collection.
   * @param item object to add to collection.
   */
  void add(Object item) {
    if (_items == null) {
      _items = [];
    }
    _items.add(item);
    int changeIndex = _items.length - 1;
    raiseChangedEvent(CollectionChangedEvent.CHANGETYPE_ADD,
        changeIndex, _NUM_ITEMS_CHANGED_ONE_ITEM);
  }

  /**
   * Inserts item into collection.
   * @param index Insertion index.
   * @param item object to insert into collection.
   */
  void insert(int index, Object item) {
    if (_items == null) {
      _items = [];
    }
    _items.insert(index, item);
    raiseChangedEvent(CollectionChangedEvent.CHANGETYPE_ADD,
        index, _NUM_ITEMS_CHANGED_ONE_ITEM);
  }

  /**
   * Inserts a list of items into collection.
   * @param index Insertion index.
   * @param list of item objects to insert into collection.
   */
  void insertAll(int index, List list) {
    if (_items == null) {
      _items = [];
    }
    int numItems = list.length;
    for (int i = 0; i < numItems; i++) {
      _items.insert(index + i, list[i]);
    }
    raiseChangedEvent(CollectionChangedEvent.CHANGETYPE_ADD,
        index, numItems);
  }

  /**
   * Removes an item from collection.
   * @param item Object to remove from collection.
   */
  void remove(Object item) {
    if (_items == null) {
      return;
    }
    int index = _items.indexOf(item, 0);
    if (index != -1) {
      raiseChangedEvent(CollectionChangedEvent.CHANGETYPE_REMOVING, index,
          _NUM_ITEMS_CHANGED_ONE_ITEM);
      _items.removeAt(index);
      raiseChangedEvent(CollectionChangedEvent.CHANGETYPE_REMOVE, index,
          _NUM_ITEMS_CHANGED_ONE_ITEM);
    }
  }

  /**
   * Returns index of item in collection.
   * @param item Object to search for.
   * @param fromIndex Search start location.
   * @return return index of object in collection. Returns -1 if item is not
   *  in the collection.
   */
  int indexOf(Object item) {
    return _items != null ? _items.indexOf(item, 0) : -1;
  }

  /**
   * Removes an item at an index from collection.
   * @param index Index of object to be remove from collection.
   */
  void removeAt(int index) {
    if (_items == null) {
      return;
    }
    raiseChangedEvent(CollectionChangedEvent.CHANGETYPE_REMOVING, index,
        _NUM_ITEMS_CHANGED_ONE_ITEM);
    _items.removeAt(index);
    raiseChangedEvent(CollectionChangedEvent.CHANGETYPE_REMOVE, index,
        _NUM_ITEMS_CHANGED_ONE_ITEM);
  }

  /**
   * Removes all items from collection.
   */
  void clear() {
    if ((_items != null) && (_items.length != 0)) {
      int count = _items.length;
      raiseChangedEvent(CollectionChangedEvent.CHANGETYPE_REMOVING,
          _START_CHANGE_INDEX, count);
      _items.clear();
      raiseChangedEvent(CollectionChangedEvent.CHANGETYPE_REMOVE,
          _START_CHANGE_INDEX, count);
    }
  }

  /**
   * Returns an item at specified index or undefined if index is out of
   * bounds.
   */
  Object getItemAt(int index) => _items[index];

  void raiseChangedEvent(int type, int index, int count) {
    CollectionChangedEvent event = CollectionChangedEvent.create(
        this, type, index, count);
    notifyListeners(changedEvent, event);
    CollectionChangedEvent.release(event);
  }

  /**
   * Copies collection contents to an array. If array already has items,
   * contents are appended to end of array.
   * @param targetArray Array to copy items to.
   */
  void copyTo(List<Object> targetArray) {
    for (int index = 0; index < _items.length; index++) {
      targetArray.add(_items[index]);
    }
  }
}
